pkg/crds/crds.go

```diff
@@ -0,0 +1,205 @@
+// Package crds is used for installing rancher CRDs.
+package crds
+
+import (
+	"context"
+	"embed"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/rancher/rancher/pkg/fleet"
+	"github.com/rancher/wrangler/pkg/crd"
+	"github.com/rancher/wrangler/pkg/yaml"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	// yamlDir directory holding CRDs yaml files.
+	yamlDir = "yamls"
+
+	// readyDuration time to wait for CRDs to be ready.
+	readyDuration = time.Minute * 1
+
+	k8sManagedByKey = "app.kubernetes.io/managed-by"
+	managerValue    = "rancher"
+)
+
+//go:embed yamls/*
+var crdFS embed.FS
+
+// EnsureRequired will ensure all required CRDs needed by Rancher based on the currently enabled features are installed and up to date.
+func EnsureRequired(ctx context.Context, cfg *rest.Config) error {
+	return Ensure(ctx, cfg, RequiredCRDs())
+}
+
+// Ensure will ensure all given CRDs are installed and up to date.
+func Ensure(ctx context.Context, cfg *rest.Config, crdNames []string) error {
+	toCreateCRDs, err := getCRDs(crdNames)
+	if err != nil {
+		return fmt.Errorf("failed to get CRDs: %w", err)
+	}
+
+	ownedByRancher, err := labels.NewRequirement(k8sManagedByKey, selection.Equals, []string{managerValue})
+	if err != nil {
+		return fmt.Errorf("failed to create crd label selector: %w", err)
+	}
+	selector := labels.NewSelector().Add(*ownedByRancher)
+
+	clientSet, err := clientset.NewForConfig(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to create new clientset: %w", err)
+	}
+
+	err = crd.BatchCreateCRDs(ctx, clientSet.ApiextensionsV1().CustomResourceDefinitions(), selector, readyDuration, toCreateCRDs)
+	if err != nil {
+		return fmt.Errorf("failed to create CRDs: %w", err)
+	}
+
+	return nil
+}
+
+// getCRDs finds all embedded CRDs that match the given list of CRDs. If a CRD is specified and not found an error is returned.
+func getCRDs(crdNames []string) ([]*apiextv1.CustomResourceDefinition, error) {
+	// create a map of crds that we want to return from this function.
+	toReturn := map[string]bool{}
+	for _, crdName := range crdNames {
+		// only return CRDs that are specified and have been migrated to the new CRD generation flow
+		if MigratedResources[crdName] {
+			toReturn[crdName] = true
+		}
+	}
+
+	allCRDs, err := crdsFromDir(yamlDir)
+	if err != nil {
+		return nil, err
+	}
+
+	capiMap := toMap(CAPICRDs())
+	bootstrapFleetMap := toMap(bootstrapFleet())
+
+	retCRDs := []*apiextv1.CustomResourceDefinition{}
+
+	for _, crd := range allCRDs {
+		if !toReturn[crd.Name] {
```

Comment:
Nit: It might make sense to return an error if we encounter a duplicate CRD here, because the definition that we will actually use is ambiguous (i.e. which one did we use and how are they different?)